cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -3356,6 +3356,17 @@
 (rule (lower (return_call_indirect sig_ref callee args))
       (gen_return_call_indirect sig_ref callee args))
 
+;; Rules for `stack_switch` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(rule (lower (stack_switch store_context_ptr load_context_ptr in_payload0))
+       ;; For the time being, stack switching is only supported on x64 Linux
+       (if (on_linux))
```

Comment:
There is no other place in Cranelift that looks at the target OS when determining how to lower clif ir. Even TLS handling chooses the right variant for the target OS using a codegen flag rather than by looking at the OS in the triple. Using the right calling conventions is done by the producer of the clif ir.

Reply:
I think I agree here: philosophically, this is if not quite literally calling-convention related, certainly like an OS-interface detail that should be "baked into" the CLIF as CLIF is ordinarily explicit about such details. Can we check the triple in the Wasm translation (i.e., the wasmtime `FuncEnvironment` hook called by `cranelift-wasm`, or wherever else this instruction is generated) and fail if on the wrong platform?

This is also a little more future-looking in the sense that there may be *Wasmtime* details at some future point related to stack-switching (e.g., what if we add our own stack-protection mitigations or have some custom kind of stack-growth scheme or ...) -- we wouldn't want to hardcode that into Cranelift lowering in the same way OS details are here.

Reply:
Just to recap the context sprinkled through this PR: The reason why we need to do something platform/OS-specific here eventually is that Windows requires us to update parts of the [Thread Information Block](https://en.wikipedia.org/wiki/Win32_Thread_Information_Block) when switching stacks. Unfortunately, what exactly needs to be done is undocumented, so I've not done it, yet. The "plain" stack switching I implemented in this PR was supposed to work only on Linux, but from what I can tell should work on x64 macOS, too.

@bjorn3:

Following what you said, I guess one approach would be to make the lowering of `stack_switch` more similar to that of `tls_value`. I'd imagine this would look as follows:
- Create a `StackSwitchMode` enum with `Default` and  `UpdateTib` variants.
- Encode a value of that type in `shared_settings::Flags` and also add it as a field to the `StackSwitch` `MInst`.
- When emitting code for `StackSwitch`, check the value of the `StackSwitchMode` flag and act accordingly.

I'm also happy to have more than one `MInst` for stack switching, and lower the `stack_switch` CLIF instruction to one of them, based on the value of `StackSwitchMode`. That would mirror `tls_value` more closely, but I'm inclined to avoid that: The core stack switching code is always the same, we just sometimes need to emit some extra code on top of that.

Finally, what I described above would be the medium-term solution once I've implemented Windows support. For the time being I would just add a `None` variant to `StackSwitchMode`, use that on Windows, and panic if we see it when emitting code for `StackSwitch`.

@cfallin:

> Can we check the triple in the Wasm translation (i.e., the wasmtime `FuncEnvironment` hook called by `cranelift-wasm`, or wherever else this instruction is generated) and fail if on the wrong platform?

Did you mean for this to be just a solution for the current issue of me wanting to fail if not on Linux? Or did you mean to also use this approach (i.e., resolving differences at the Wasm -> CLIF stage) in the future when we want to generate slightly different code on different OS-es?

In the latter case, are you suggesting to have multiple stack switching CLIF instructions, for the case with and without the TIB update, then choose between them during the Wasm translation? Or are you suggesting to have a single CLIF instruction for the stack switching itself, but then emit some additional CLIF on Windows (say, some additional loads and stores) around the `stack_switch` during the Wasm translation?

I fully agree that for things like stack growing there's a good chance that in the future, we'd need some more customization of the generated code (e.g., customize stack probing, function preludes, ...).

Reply:
> Did you mean for this to be just a solution for the current issue of me wanting to fail if not on Linux? Or did you mean to also use this approach (i.e., resolving differences at the Wasm -> CLIF stage) in the future when we want to generate slightly different code on different OS-es?

Both, I think -- basically, any sort of difference of behavior on that level should be reified in the CLIF, rather than implicit; that's how we've handled things like TLS and other platform dependencies.

> In the latter case, are you suggesting to have multiple stack switching CLIF instructions, for the case with and without the TIB update, then choose between them during the Wasm translation? Or are you suggesting to have a single CLIF instruction for the stack switching itself, but then emit some additional CLIF on Windows (say, some additional loads and stores) around the `stack_switch` during the Wasm translation?

One or the other, depending on what is actually needed? I don't have enough context to actually specify the full design; that's something we can discuss further; only that we should make it explicit *somehow*. If one platform requires a superset of the work that another platform does, factoring out the common bit as one instruction and adding more logic at the CLIF level seems reasonable. On the other hand it's totally reasonable to have `stack_switch_windows` and `stack_switch_sysv` instructions IMHO; again see how we did TLS, with separate instructions for ELF and Mach-O cases.

Reply:
> On the other hand it's totally reasonable to have `stack_switch_windows` and `stack_switch_sysv` instructions IMHO; again see how we did TLS, with separate instructions for ELF and Mach-O cases.

Ah, I think that's where my confusion came from: For TLS, there is a single CLIF instruction, which is then lowered one of  several per-platform `MInst`s (but based on a flag in the backend, not some ad hoc OS check like I did). I'm happy to do it like that, which would be similar to what I described in my response to @bjorn3.

I'd prefer that over moving the TIB update out of the `stack_switch` (CLIF) instruction itself.